<a href="https://colab.research.google.com/github/ShymaShameer/Eniac_project/blob/main/Project_IDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project**
**Case Study: Eniac’s Discount Strategy**


# Importing library

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

# Loading Data

In [ ]:
url = 'https://drive.google.com/file/d/1m1ThDDIYRTTii-rqM5SEQjJ8McidJskD/view?usp=drive_link'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
brands_df_original = pd.read_csv(path)

In [ ]:
url = 'https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=drive_link'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
orderlines_df_original = pd.read_csv(path)

In [ ]:
url = 'https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=drive_link'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
orders_df_original = pd.read_csv(path)

In [ ]:
url = 'https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=drive_link'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
products_df_original = pd.read_csv(path)

In [ ]:
# making a copy of all dataframes

brands_df = brands_df_original.copy()
orderlines_df = orderlines_df_original.copy()
orders_df = orders_df_original.copy()
products_df = products_df_original.copy()

# Initial Data Exploration

## Brands

In [ ]:
brands_df.head()

,short,long
0,8MO,8Mobility
1,ACM,Acme
2,ADN,Adonit
3,AII,Aiino
4,AKI,Akitio


In [ ]:
brands_df.shape

print(f"The number of rows is: {brands_df.shape[0]}")
print(f"The number of columns is: {brands_df.shape[1]}")

The number of rows is: 187
The number of columns is: 2


In [ ]:
brands_df.size

374

In [ ]:
brands_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187 entries, 0 to 186
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   short   187 non-null    object
 1   long    187 non-null    object
dtypes: object(2)
memory usage: 3.1+ KB


In [ ]:
brands_df.describe()

,short,long
count,187,187
unique,187,181
top,8MO,Apple
freq,1,2


In [ ]:
brands_df.nunique()

,0
short,187
long,181


In [ ]:
brands_df.isna().sum()

,0
short,0
long,0


In [ ]:
brands_df.duplicated().sum()

np.int64(0)

In [ ]:
# for identifying any pattern:
# a) string length in "short" ( All strings are of 3 & alphanumeric)

#brands_df["short"].str.len()
brands_df["short"].str.len().nunique()

1

In [ ]:
# b) string length in "long"  (String length varies from 2- 16 & alhphanumeric)
#brands_df["long"].str.len()
brands_df["long"].str.len().unique()

array([ 9,  4,  6,  5, 10, 16,  8,  7,  3, 13, 12, 11, 15,  2])

## Orders

In [ ]:
orders_df.head(5)

,order_id,created_date,total_paid,state
0,241319,2017-01-02 13:35:40,44.99,Cancelled
1,241423,2017-11-06 13:10:02,136.15,Completed
2,242832,2017-12-31 17:40:03,15.76,Completed
3,243330,2017-02-16 10:59:38,84.98,Completed
4,243784,2017-11-24 13:35:19,157.86,Cancelled


In [ ]:
orders_df.shape

print(f"The number of rows is: {orders_df.shape[0]}")
print(f"The number of columns is: {orders_df.shape[1]}")

The number of rows is: 226909
The number of columns is: 4


In [ ]:
orders_df.size

907636

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


In [ ]:
orders_df.describe().round(2)

,order_id,total_paid
count,226909.00,226904.00
mean,413296.48,569.23
std,65919.25,1761.78
min,241319.00,0.00
25%,356263.00,34.19
50%,413040.00,112.99
75%,470553.00,525.98
max,527401.00,214747.53


In [ ]:
# to identify difference in distribution pattern
orders_df["total_paid"].agg(['mean', 'median']).round(2)

,total_paid
mean,569.23
median,112.99


In [ ]:
orders_df.nunique()

,0
order_id,226909
created_date,224828
total_paid,31236
state,5


In [ ]:
orders_df.isna().sum()

,0
order_id,0
created_date,0
total_paid,5
state,0


In [ ]:
orders_df.duplicated().sum()

np.int64(0)

In [ ]:
# for identifying any pattern

#orders_df.loc[orders_df["total_paid"]==0,:]
#orders_df["total_paid"].nunique()
orders_df["total_paid"].value_counts().sort_values(ascending= False)

,count
total_paid,
0.00,19334
19.99,2812
29.99,2193
39.99,1504
49.99,1495
...,...
1311.92,1
66.55,1
158.83,1


In [ ]:
#
completed_orders = orders_df.loc[(orders_df["state"] == "Completed") & (orders_df["total_paid"] == 0.0), :]
completed_orders

,order_id,created_date,total_paid,state
150,296010,2017-01-09 23:47:00,0.0,Completed
2061,301495,2017-01-03 12:22:02,0.0,Completed
30427,329911,2017-02-27 18:21:01,0.0,Completed
61224,360765,2017-05-30 23:28:57,0.0,Completed
61902,361443,2017-06-01 23:21:54,0.0,Completed
61977,361518,2017-06-02 06:18:27,0.0,Completed
63610,363153,2017-06-07 13:39:13,0.0,Completed
65940,365485,2017-06-14 10:37:02,0.0,Completed
82898,382473,2017-07-26 18:23:09,0.0,Completed
82938,382513,2017-07-26 19:23:21,0.0,Completed


**Observation & action plans:**

* Created_date : should be date format
* Total_paid:  
1. missing values
2. Mean-median gap is high (569.23- 112.99), so extreme high value outliers. Also 75% of orders are for 525.98 or less, while max value is 214,747.53
3. 0.00 value for 19334 orders


## Orderlines

In [ ]:
orderlines_df.head()

,id,id_order,product_id,product_quantity,sku,unit_price,date
0,1119109,299539,0,1,OTT0133,18.99,2017-01-01 00:07:19
1,1119110,299540,0,1,LGE0043,399.00,2017-01-01 00:19:45
2,1119111,299541,0,1,PAR0071,474.05,2017-01-01 00:20:57
3,1119112,299542,0,1,WDT0315,68.39,2017-01-01 00:51:40
4,1119113,299543,0,1,JBL0104,23.74,2017-01-01 01:06:38


In [ ]:
orderlines_df.shape

print(f"The number of rows is: {orderlines_df.shape[0]}")
print(f"The number of columns is: {orderlines_df.shape[1]}")

The number of rows is: 293983
The number of columns is: 7


In [ ]:
orderlines_df.size

2057881

In [ ]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


In [ ]:
orderlines_df.describe().round(2)

,id,id_order,product_id,product_quantity
count,293983.00,293983.00,293983.0,293983.00
mean,1397918.10,419999.12,0.0,1.12
std,153009.56,66344.49,0.0,3.40
min,1119109.00,241319.00,0.0,1.00
25%,1262542.50,362258.50,0.0,1.00
50%,1406940.00,425956.00,0.0,1.00
75%,1531321.50,478657.00,0.0,1.00
max,1650203.00,527401.00,0.0,999.00


In [ ]:
orderlines_df.nunique()

,0
id,293983
id_order,204855
product_id,1
product_quantity,67
sku,7951
unit_price,11329
date,251631


In [ ]:
orderlines_df.isna().sum()

,0
id,0
id_order,0
product_id,0
product_quantity,0
sku,0
unit_price,0
date,0


In [ ]:
orderlines_df.duplicated().sum()

np.int64(0)

In [ ]:
# for identifying any pattern
# a) sku: pattern- majority 7 strings (5 with 11 strings, 295  with 5 str, 11646 with 9 str)

#orderlines_df["sku"].str.len().nunique()
orderlines_df["sku"].str.len().value_counts()

,count
sku,
7,282037
9,11646
5,295
11,5


In [ ]:
# b) product_id
#orderlines_df["product_id"].nunique()
orderlines_df["product_id"].value_counts()

,count
product_id,
0,293983


**Observations & action plan:**
* Product_id: not in use now, only 1 unique value(0), can be deleted
* Product_quantity: max -999 (seems to be outlier since mean is 1.12, also 75% of orders are 1). Can explore more
* Unit_price: should be float
* Date: should be date format

## Products

In [ ]:
products_df.head()

,sku,name,desc,price,promo_price,in_stock,type
0,RAI0007,Silver Rain Design mStand Support,Aluminum support compatible with all MacBook,59.99,499.899,1,8696
1,APP0023,Apple Mac Keyboard Keypad Spanish,USB ultrathin keyboard Apple Mac Spanish.,59,589.996,0,13855401
2,APP0025,Mighty Mouse Apple Mouse for Mac,mouse Apple USB cable.,59,569.898,0,1387
3,APP0072,Apple Dock to USB Cable iPhone and iPod white,IPhone dock and USB Cable Apple iPod.,25,229.997,0,1230
4,KIN0007,Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM,2GB RAM Mac mini and iMac (2006/07) MacBook Pr...,34.99,31.99,1,1364


In [ ]:
products_df.shape

print(f"The number of rows is: {products_df.shape[0]}")
print(f"The number of columns is: {products_df.shape[1]}")

The number of rows is: 19326
The number of columns is: 7


In [ ]:
products_df.size

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


In [ ]:
products_df.describe().round(2)

,in_stock
count,19326.00
mean,0.11
std,0.31
min,0.00
25%,0.00
50%,0.00
75%,0.00
max,1.00


In [ ]:
products_df.nunique()

,0
sku,10579
name,10373
desc,7098
price,2690
promo_price,4614
in_stock,2
type,126


In [ ]:
products_df.isna().sum()

,0
sku,0
name,0
desc,7
price,46
promo_price,0
in_stock,0
type,50


In [ ]:
products_df.duplicated().sum()
print(f"The number of duplicated rows is: {products_df.duplicated().sum()}")

The number of duplicated rows is: 8746


In [ ]:
products_df.loc[products_df.duplicated, :]

,sku,name,desc,price,promo_price,in_stock,type
101,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
102,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
103,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
104,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
105,APP0390,"Apple MacBook Pro 133 ""Core i5 25GHz | 4GB RAM...",MacBook Pro laptop 133 inches (MD101Y / A).,1199,11.455.917,0,1282
...,...,...,...,...,...,...,...
16831,APP2302,"Apple MacBook Pro 13 ""Core i5 Touch Bar 33GHz ...",New MacBook Pro 13-inch Core i5 Touch Bar 33 G...,26.155.941,26.155.941,0,"1,02E+12"
16833,APP2303,"Apple MacBook Pro 13 ""Core i5 Touch Bar 33GHz ...",New MacBook Pro 13 inch Touch Bar 33 GHz Core ...,237.559.421,23.755.942,0,"1,02E+12"
18190,PAR0077,Parrot Bebop Drone 2 Power,Drone cuadricóptero quality camera integrated ...,699.9,6.733.892,0,11905404
18308,NKI0010,Nokia Wireless sphygmomanometer Plata,Sphygmomanometer for iPhone iPad and iPod App.,129.99,1.149.899,1,11905404


In [ ]:
# for identifying any pattern
# a) desc
products_df.loc[products_df["desc"].isna(),:]

,sku,name,desc,price,promo_price,in_stock,type
16126,WDT0211-A,"Open - Purple 2TB WD 35 ""PC Security Mac hard ...",NaN,107,814.659,0,1298
16128,APP1622-A,Open - Apple Smart Keyboard Pro Keyboard Folio...,NaN,1.568.206,1.568.206,0,1298
17843,PAC2334,Synology DS718 + NAS Server | 10GB RAM,NaN,566.35,5.659.896,0,12175397
18152,KAN0034-A,Open - Kanex USB-C Gigabit Ethernet Adapter Ma...,NaN,29.99,237.925,0,1298
18490,HTE0025,Hyper Pearl 1600mAh battery Mini USB Mirror an...,NaN,24.99,22.99,1,1515
18612,OTT0200,OtterBox External Battery Power Pack 20000 mAHr,NaN,79.99,56.99,1,1515
18690,HOW0001-A,Open - Honeywell thermostat Lyric zonificador ...,NaN,199.99,1.441.174,0,11905404


**Observations & Action plan:**
* The number of duplicated rows is: 8746
* Price:
1. should be float
2. missing values
* Promo_price: should be float
* Desc_ missing value
* Type:
1.missing value
2.Type should be numerical as per the Data info
* In_stock:  values of 0 & 1 to indicate if present in stock or not , can turn to str values if needed for better visualisation.


# Custom Exploration

* How many orders are there?
* How many products are there?
* What period of time do these orders comprise?
* How many orders are Completed?
* How should revenue be computed?




In [ ]:
# How many orders are there? : 226909
orders_df.nunique()

In [ ]:
# How many products are there?  :10579
products_df.nunique()

In [ ]:
# What period of time do these orders comprise?: 01-01-2017 to 14-03-2018 (437 days)
# orderlines_df["date"].max()
# orderlines_df["date"].min()


In [ ]:
import datetime
max_date= datetime.datetime(2018,3,14)
min_date= datetime.datetime(2017,1,1)
date_diff = max_date - min_date

print(f"First order date is: {min_date} \n Last order date is: {max_date}")
print(date_diff)

First order date is: 2017-01-01 00:00:00 
 Last order date is: 2018-03-14 00:00:00
437 days, 0:00:00


In [ ]:
#How many orders are Completed?: 46605
orders_df.groupby("state")["order_id"].count()

In [ ]:
# How should revenue be computed?